### PR TITLE
Istio enable external

### PR DESCRIPTION
**Pull request by ganadurai**
## Description
Enabling the tool to setup ingressgateway so the httpbin service is exposed externally and the incoming request is protected by the Envoy proxy sidecar with apigee adapter.

## Housekeeping
(please check all that apply [x], do not edit the text)
- [x] I have run all the tests locally and they all pass.
- [x] I have followed the relevant style guide for my changes.
- [ ] PR requires full pipeline run (Run for changes only by default).

**CC:** @apigee-devrel-reviewers

### CODE REVIEW

**Comment by OmidTahouri** [comment]
edited PR description to not require a full pipeline run

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+1. Deployment of starter  setup of apigee-envoy deployment within Istio enabled Kubernetes platform. 
+
+### Installation:
+
+1. **Set your GCP Project ID, Apigee platform environment variables.** 
+    ```bash
+    export PROJECT_ID=<your-project-id>
+    export CLUSTER_NAME=<gke-cluster-name>
+    export CLUSTER_LOCATION=<gke-cluster-region>
+    export APIGEE_PROJECT_ID="x-project-1-344916"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:22]: can we make this a `<gcp-projec>` variable like the ones above

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Apigee Envoy extension (for external access)
+
+This is an extension to the starter setup of apigee-envoy deployment within GKE container platform with Apigee X/Hybrid as API management platform. 
+
+The starter kit deploys httpbin service configured with Envoy proxies as side car proxies. The Envoy proxy is enabled with apigee-adapter as step in the request path enforcing  Apigee provided authentication methods.
+
+This extension enables the expose of deployed sample application (httpbin) externally via istio-ingressgateway. It showcases apigee enabled envoy proxies can offer protection of api traffic intiated externally.
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:7]: * the expose => the exposure?
* apigee => Apigee
* api => API
* ? externally => outside the mesh

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    gcloud auth application-default login --no-launch-browser
+
+    export TOKEN=$(gcloud auth print-access-token)
+    ```
+
+1. **Download the Apigee Envoy PoC Toolkit binary. If already present during the apigee-envoy setup, this step can be skipped** 
+    ```bash
+    mkdir apigee-envoy-toolkit && cd "$_"
+    export ENVOY_HOME=$(pwd)
+    wget -O devrel.zip https://github.com/apigee/devrel/archive/refs/heads/main.zip
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:37]: Why don't we just git clone devrel? one step stead of 4?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+echo curl -i http://"$EXTERNAL_IP"/headers -H "\"x-api-key: $CONSUMER_KEY\"" -H "\"Host: httpbin.apigee.svc.cluster.local\""
+echo "Waiting for the deployments to be complete (35s)..."
+
+sleep 35;
+testExternalAccess;
+RESULT=$?
+
+counter=0;
+while [ $RESULT -ne 0 ] && [ $counter -lt 5 ]; do
+  printf "\n\nTesting the httpbin applicatio via external access %s of 5\n" "$counter"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:202]: application

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        port:
+          number: 80
+EOF
+
+kubectl apply -n $ASM_GATEWAYS_NAMESPACE \
+  -f "$ENVOY_HOME"/httpbin-virtual-service.yaml
+
+fetchConsumerKey;
+
+echo "Testing httpbin via externalip..."
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:192]: external IP

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    - destination:
+        host: httpbin.apigee.svc.cluster.local
+        port:
+          number: 80
+EOF
+
+kubectl apply -n $ASM_GATEWAYS_NAMESPACE \
+  -f "$ENVOY_HOME"/httpbin-virtual-service.yaml
+
+fetchConsumerKey;
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:190]: consider writing it as:

```sh
CONSUMER_KEY=$(fetchConsumerKey)
```

where the `fetchConsumerKey` function writes the key to stdout. And the error to stderr.

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  counter=$((counter+1))
+done
+
+if [[ $counter != 5 ]]; then
+  printf "\n"
+  echo "SUCCESS: use the above curl command to validate external access of deployed httpbin service protected by envoy-apigee integration"
+  printf "\n"
+fi
+
+echo "Note: The reason for host override is bcos of the Apigee product is protecting this api host. \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:215]: because this product is protected by Apigee.

Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:215]: Maybe we'll want to put this into the README?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+The request hitting the Loadbalancer (via istio-ingressgateway) can carry any valid domain name resolving to the LB. \
+The apigee product protects the k8s ClusterIP service host and hence needed the Host override in the request. 
+Only for this service host the envoy apigee adapter will trigger the authn with Apigee runtime."
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:218]: apigee => Apigee

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+printf "\n"
+echo "If a domain name is alias'd to the LB and the request should be protected via Envoy, this domain name should be added to the Apigee product."
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:220]: Maybe needs more explanation in the README

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+installIstioIngressGateway() {
+
+  git clone https://github.com/GoogleCloudPlatform/anthos-service-mesh-packages.git \
+  "$ENVOY_HOME"/asm
+
+  kubectl create namespace $ASM_GATEWAYS_NAMESPACE
+
+  kubectl label namespace $ASM_GATEWAYS_NAMESPACE istio-injection=enabled istio.io/rev-
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:78]: this label looks incomplete

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+getIngressGatewayExternalIP() {
+  EXTERNAL_IP=$(kubectl --context="${CLUSTER_CTX}" get svc -n "$ASM_GATEWAYS_NAMESPACE" -o json | \
+  jq '.items[0].status.loadBalancer.ingress[0].ip' | \
+  cut -d '"' -f 2)
+  export EXTERNAL_IP;
+}
+
+installIstioIngressGateway() {
+
+  git clone https://github.com/GoogleCloudPlatform/anthos-service-mesh-packages.git \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:73]: this is a bit fragile. Can we pin a specific release?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+  kubectl create namespace $ASM_GATEWAYS_NAMESPACE
+
+  kubectl label namespace $ASM_GATEWAYS_NAMESPACE istio-injection=enabled istio.io/rev-
+
+  kubectl apply -n $ASM_GATEWAYS_NAMESPACE \
+    -f "$ENVOY_HOME"/asm/samples/gateways/istio-ingressgateway
+
+  kubectl rollout restart deployment -n $ASM_GATEWAYS_NAMESPACE
+
+  getIngressGatewayExternalIP;
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:85]: can we change this to 

```sh
EXTERNAL_IP=$(getIngressGatewayExternalIP)
```

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+  counter=0;
+  while [[ "$EXTERNAL_IP" == 'null' ]] && [[ $counter -lt 10 ]]; do
+    printf "\n\nTesting the availability of EXTERNAL_IP for ingressgateway,  %s of 10\n" "$counter"
+    sleep 5
+    getIngressGatewayExternalIP;
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:91]: ditto

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+ISTO_INGRESSGATEWAY_CNT=$(kubectl --context="${CLUSTER_CTX}" get svc -n "$ASM_GATEWAYS_NAMESPACE" -o json | \
+jq '.items|length')
+
+if [ "$ISTO_INGRESSGATEWAY_CNT" == 0 ]; then
+  echo "No istio ingressgateway found, installing..."
+  installIstioIngressGateway;
+fi
+
+echo "Getting ingressgateway external ipaddress..."
+getIngressGatewayExternalIP;
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:138]: ditto

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+# limitations under the License.
+
+set -e
+
+init() {
+  export CLUSTER_CTX="gke_${PROJECT_ID}_${CLUSTER_LOCATION}_${CLUSTER_NAME}"
+  export ASM_GATEWAYS_NAMESPACE="istio-ingressgateway"
+  export USE_GKE_GCLOUD_AUTH_PLUGIN=True
+  export TOKEN_TYPE="Bearer"
+  export MGMT_HOST="https://apigee.googleapis.com"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:24]: Variable is used only once and never meant to change? Maybe just put the hostname directly in the curl?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+init() {
+  export CLUSTER_CTX="gke_${PROJECT_ID}_${CLUSTER_LOCATION}_${CLUSTER_NAME}"
+  export ASM_GATEWAYS_NAMESPACE="istio-ingressgateway"
+  export USE_GKE_GCLOUD_AUTH_PLUGIN=True
+  export TOKEN_TYPE="Bearer"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:23]: Used only once and the only option. Put it directly in curl?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+  return $RESULT
+}
+
+echo 'Initializing...'
+init;
+
+echo 'Validating...'
+validate;
+
+ISTO_INGRESSGATEWAY_CNT=$(kubectl --context="${CLUSTER_CTX}" get svc -n "$ASM_GATEWAYS_NAMESPACE" -o json | \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:129]: This technically only checks if there are any services in the gateway namespace? More defensive would maybe be to check for specific labels on the service?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+echo "Getting ingressgateway external ipaddress..."
+getIngressGatewayExternalIP;
+
+if [[ -z "$EXTERNAL_IP" ]] || [[ "$EXTERNAL_IP" == 'null' ]]; then
+    echo "ExternalIP is not set on the istio ingressgateway, exiting"
+    exit 1
+fi
+
+kubectl --context="${CLUSTER_CTX}" get svc -n "$ASM_GATEWAYS_NAMESPACE"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:145]: This is gonna be quite verbose in the output. is this just for debugging or absolutely needed?

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+ servers:
+   - port:
+       number: 80
+       name: http
+       protocol: HTTP
+     hosts:
+       - "*"
+EOF
+
+kubectl apply -n $ASM_GATEWAYS_NAMESPACE \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:165]: ditto

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+spec:
+ selector:
+   istio: ingressgateway
+ servers:
+   - port:
+       number: 80
+       name: http
+       protocol: HTTP
+     hosts:
+       - "*"
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/setup-external-access.sh:162]: What if you have multiple services you want to expose?

**Review by danistrebel — CHANGES_REQUESTED** [review]

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh (hunk 20 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/setup-external-access.sh:162]: As this is just a starter kit showcasing the ingressgateway setup for Apigee protected mesh services, keeping it simple on the LB front. Making the ingress accessible directly with the external ip address and hence wildcard on the Virtualservice defn.

File: tools/apigee-envoy-quickstart/extension/setup-external-access.sh (hunk 17 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/setup-external-access.sh:129]: Good catch. I should be checking if the mesh already has an ingressgateway. If so, I should be using it. Let me adjust the code to accommodate that scenario

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+     - httpbin-gateway
+     http:
+     - route:
+       - destination:
+           host: $TARGET_SERVICE_NAME.$TARGET_SERVICE_NAMESPACE.svc.cluster.local
+           port:
+             number: 80
+   EOF
+
+   envsubst < /tmp/httpbin-virtual-service.tmpl > /tmp/httpbin-virtual-service.yaml
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:131]: you could get the variables directly in the cat. no need for cat + envsubst

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    CONSUMER_KEY=$(curl -H "Authorization: Bearer ${TOKEN}"  \
+    -H "Content-Type:application/json" \
+    "https://apigee.googleapis.com/v1/organizations/${APIGEE_ORG}/developers/test-envoy@google.com/apps/envoy-adapter-app-2" | \
+    jq '.credentials[0].consumerKey'); \
+    CONSUMER_KEY=$(echo "$CONSUMER_KEY"|cut -d '"' -f 2); 
+    ```
+
+1. **Extract IP address exposed via ingressgateway component**
+    ```bash
+    EXTERNAL_IP=$(kubectl --context="${CLUSTER_CTX}" get svc -n "$ISTIO_GATEWAY_NS" -o json | \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:147]: don't you have this already from above `INGRESS_HOST`

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 21 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md:131]: duh....

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 22 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md:147]: yepp... cleaning it

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+1. **Set up login profile for your cluster.**
+    ```bash
+    gcloud container clusters get-credentials "$CLUSTER_NAME" \
+     --zone "$CLUSTER_LOCATION" \
+     --project "$GKE_PROJECT_ID"
+     export CLUSTER_CTX="gke_${GKE_PROJECT_ID}_${CLUSTER_LOCATION}_${CLUSTER_NAME}"
+    ```
+
+1. ### Set the namespace hosting istio-ingressgateway
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:50]: How do we ensure that these instructions are compatible with the istio/asm version that a user has installed?

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+     servers:
+       - port:
+           number: 80
+           name: http
+           protocol: HTTP
+         hosts:
+           - "$TARGET_HOST"
+   EOF
+
+   kubectl apply -n $ISTIO_GATEWAY_NS -f /tmp/httpbin-gateway.yaml
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:106]: Can we pipe the template output above directly into kubectl? Fewer files dangling in /tmp

```
cat <<EOF | kubectl apply -f -
...
EOF
```

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+     - httpbin-gateway
+     http:
+     - route:
+       - destination:
+           host: $TARGET_SERVICE_NAME.$TARGET_SERVICE_NAMESPACE.svc.cluster.local
+           port:
+             number: 80
+   EOF
+
+   kubectl apply -n $ISTIO_GATEWAY_NS -f /tmp/httpbin-virtual-service.yaml
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:129]: ditto

File: tools/apigee-envoy-quickstart/extension/README.md [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+    ```
+
+1. **Extract the consumer key from the developer app, this is created on Apigee platform during the [quickstart setup](https://github.com/apigee/devrel/tree/main/tools/apigee-envoy-quickstart#envoy-with-apigee-adapter-as-containers-within-kubernetes-platform) :**
+    ```bash
+    CONSUMER_KEY=$(curl -H "Authorization: Bearer ${TOKEN}"  \
+    -H "Content-Type:application/json" \
+    "https://apigee.googleapis.com/v1/organizations/${APIGEE_ORG}/developers/test-envoy@google.com/apps/envoy-adapter-app-2" | \
```
Comment (danistrebel) [inline tools/apigee-envoy-quickstart/extension/README.md:136]: The naming is a bit ugly any reason for the -2 suffix?

Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md:136]: Just a carry over from the initial set up days, nothing significant.

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 23 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md:49]: which of these commands are specific to a version for istio/asm ? May be the the tool should set the expected platform whats its coded/tested for and ask users to modify it for their respective versions? As this is just a 'quickstart' tool.

Thoughts?

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 24 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md]: fixed.

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 25 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/extension/README.md]: fixed

File: tools/apigee-envoy-quickstart/README.md [hunk 27]
```diff
@@ -79,7 +79,7 @@ The Apigee Envoy Quickstart Toolkit sets up the Envoy proxies with Apigee adapte
 ### Cleanup
 1. **Run to cleanup the PoC setup from the GKE and Apigee platform**
     ```bash
-    cd ${ENVOY_HOME}/apigee-envoy-quickstart-main
+    cd ${ENVOY_HOME}/apigee-devrel/tools/apigee-envoy-quickstart/
```
Comment (OmidTahouri) [inline tools/apigee-envoy-quickstart/README.md:82]: I'm a little confused by this. Is the entire devrel repo cloned inside of `$ENVOY_HOME`?

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 23 above)
Comment (OmidTahouri) [reply tools/apigee-envoy-quickstart/extension/README.md:49]: I can't see any Istio or ASM specific commands, but I guess it's just a question of how compatible this tool should be. On this occassion, @ganadurai makes a valid point that this is all in the context of a highly opinionated "quickstart" which gets you up and running quickly, but potentially has limited flexibility and compatibility.

It's likely @danistrebel is actually referring to the *version* of Istio and not the distribution (e.g. ASM). I'm not sure what guarantees can be put in place for that. Perhaps the scripts could do a check of the Istio version currently installed and ensure it's at `> x.x.x` 🤔

File: tools/apigee-envoy-quickstart/README.md (hunk 27 above)
Comment (ganadurai) [reply tools/apigee-envoy-quickstart/README.md:82]: Yes. entire devrel is cloned and executed from within apigee-envoy-quickstart directory.

File: tools/apigee-envoy-quickstart/extension/README.md (hunk 23 above)
Comment (danistrebel) [reply tools/apigee-envoy-quickstart/extension/README.md:49]: just a suggestion. Happy to leave it as is.

**Review by danistrebel — APPROVED** [review]

**Pull request merged**